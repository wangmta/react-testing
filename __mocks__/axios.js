const axiosMock = jest.genMockFromModule('axios');

let mockResponse = {
  data: {
    stores: [{ location: 'test location', address: 'test address' }]
  }
};

axiosMock.get.mockImplementation(req);

function req() {
  return new Promise((resolve, reject) => {
    axiosMock.delayTimer = setTimeout(() => {
      resolve(mockResponse);
    }, 100);
  });
}

module.exports = axiosMock;
