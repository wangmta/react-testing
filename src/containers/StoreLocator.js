import React, { Component } from 'react';
import Button from '../components/Button';
import Map from '../components/Map';
import mapChooser from '../mapChooser';
import axios from 'axios';

class StoreLocator extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentMap: 'none.png',
      stores: []
    };
    this.chooseMap = this.chooseMap.bind(this);
  }

  async componentDidMount() {
    let response = await axios.get('http://localhost:3000/data/stores.json');
    this.setState({ stores: response.data.stores });
  }

  chooseMap(event) {
    this.setState({ currentMap: mapChooser(event.target.value) });
  }

  render() {
    let storeButtons = this.state.stores.map((store, id) => {
      return <Button key={id} location={store.location} handleClick={this.chooseMap} />;
    });
    return (
      <div>
        <div>
          {/* <Button location="Portland" /> <Button location="Astoria" /> <Button /> */}
          {storeButtons}
        </div>
        <Map imagename={this.state.currentMap} location={this.props.location} />
      </div>
    );
  }
}

export default StoreLocator;
