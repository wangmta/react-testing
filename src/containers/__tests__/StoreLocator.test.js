// smoke test, test if the component can render without crashing
import React from 'react';
import StoreLocator from '../StoreLocator';
import { shallow } from 'enzyme';
import axios from 'axios';
import renderer from 'react-test-renderer';

describe('StoreLocator tests', () => {
  let mountedStoreLocator;

  beforeEach(() => {
    mountedStoreLocator = shallow(<StoreLocator />);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<StoreLocator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls axios.get in #componentDidMount', () => {
    return mountedStoreLocator
      .instance()
      .componentDidMount()
      .then(() => {
        expect(axios.get).toHaveBeenCalled();
      });
  });

  it('calls axios.get with correct url', () => {
    return mountedStoreLocator
      .instance()
      .componentDidMount()
      .then(() => {
        expect(axios.get).toHaveBeenCalledWith('http://localhost:3000/data/stores.json');
      });
  });

  it('updates state with api data', () => {
    return mountedStoreLocator
      .instance()
      .componentDidMount()
      .then(() => {
        expect(mountedStoreLocator.state()).toHaveProperty('stores', [
          { location: 'test location', address: 'test address' }
        ]);
      });
  });

  it('renders 1 button after calling api', () => {
    return mountedStoreLocator
      .instance()
      .componentDidMount()
      .then(() => {
        const buttons = mountedStoreLocator.find('Button');
        expect(buttons.length).toBe(1);
      });
  });

  it('renders without crashing', () => {
    //   const div = document.createElement('div');
    //   ReactDOM.render(<StoreLocator />, div);

    // shawllow rendering: isolate the component, without rendering it's child components,
    // using enzyme npm i --save-dev enzyme enzyme-adapter-react-16
    let mountedStoreLocator = shallow(<StoreLocator />);
  });

  // it('renders 3 buttons', () => {
  //   const buttons = mountedStoreLocator.find('Button');
  //   expect(buttons.length).toBe(3);
  // });

  it('renders 1 map', () => {
    const map = mountedStoreLocator.find('Map');
    expect(map.length).toBe(1);
  });
});

describe('chooseMap', () => {
  it('updates this.state.currentMap using the location passed to it', () => {
    let mountedStoreLocator = shallow(<StoreLocator />);
    let mockEvent = { target: { value: 'testland' } };
    mountedStoreLocator.instance().chooseMap(mockEvent);
    expect(mountedStoreLocator.instance().state.currentMap).toBe('testland.png');
  });
});
