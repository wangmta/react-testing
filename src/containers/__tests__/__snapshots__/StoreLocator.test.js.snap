// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreLocator tests renders correctly 1`] = `
<div>
  <div />
  <div
    className="MapBox"
  >
    <img
      alt={undefined}
      src="images/none.png"
    />
  </div>
</div>
`;
